### PR TITLE
multi: implement wait using winsock events

### PR DESCRIPTION
**Pull request by rcombs**
Repeat of #5397, but this time with #5631 fixed. In theory, anyway; I haven't had a chance to test this out on a Windows machine myself yet. Also not sure how to add a test (particularly one for something that takes longer than it should, but otherwise finishes correctly).

### CODE REVIEW

**Comment by mback2k** [comment]
Thanks, I will take a look later this week. This reminds me of a similar issue and solution I had with `sockfilt.c`.

File: lib/multi.c [hunk 1]
```diff
@@ -1145,22 +1163,44 @@ static CURLMcode Curl_multi_wait(struct Curl_multi *multi,
...
         }
         if(bitmap & GETSOCK_WRITESOCK(i)) {
+#ifdef USE_WINSOCK
+          struct timeval timeout;
+          fd_set writefds;
+          timeout.tv_sec = 0;
+          timeout.tv_usec = 0;
+          FD_ZERO(&writefds);
+          FD_SET(sockbunch[i], &writefds);
+          if(select(1, NULL, &writefds, NULL, &timeout) == 1)
```
Comment (mback2k) [inline lib/multi.c:1187]: Shoudn't the first argument of select be `writefds + 1` (max fd + 1)? I know WinSock does not really use it, but at least in theory that should be value.

File: lib/multi.c [hunk 2]
```diff
@@ -1208,10 +1303,49 @@ static CURLMcode Curl_multi_wait(struct Curl_multi *multi,
...
+
+          for(i = 0; i < MAX_SOCKSPEREASYHANDLE; i++) {
+            if(bitmap & (GETSOCK_READSOCK(i) | GETSOCK_WRITESOCK(i))) {
+              WSANETWORKEVENTS events = {0};
+              if(WSAEnumNetworkEvents(sockbunch[i], multi->wsa_event,
+                                      &events) == 0) {
+                if(events.lNetworkEvents != 0)
+                  retcode++;
+              }
+              if(ret && !events.lNetworkEvents && GETSOCK_WRITESOCK(i)) {
```
Comment (mback2k) [inline lib/multi.c:1327]: This causes:
```
In file included from multi.c:35:
multi.c: In function 'Curl_multi_wait':
multiif.h:52:33: error: '<<' in boolean context, did you mean '<'? [-Werror=int-in-bool-context]
   52 | #define GETSOCK_WRITESOCK(x) (1 << (GETSOCK_WRITEBITSTART + (x)))
      |                              ~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
multi.c:1327:51: note: in expansion of macro 'GETSOCK_WRITESOCK'
 1327 |               if(ret && !events.lNetworkEvents && GETSOCK_WRITESOCK(i)) {
      |         
```

File: lib/multi.c (hunk 1 above)
Comment (rcombs) [reply lib/multi.c]: Oh whoops, yeah that's definitely correct, I always forget about `select`'s weird bitmask interactions. Fixed.

File: lib/multi.c (hunk 2 above)
Comment (rcombs) [reply lib/multi.c]: …bit again by another bitwise thing! Fixed.

**Comment by mback2k** [comment]
One thing which we missed last time: This makes curl itself depend on WinSock2 for the first time. Currently it is theoretically possible to still run curl on versions older than Windows 95 (NT 3.5), but with this change curl requires at least Windows 98 (NT 4.0) or Windows 95 (NT 3.5) with the WinSocks2 addon package. I don't know if these historic Windows versions still matter to us.

**Comment by mback2k** [comment]
> LGTM. I don't think requiring WinSock 2 is a problem as we require the February 2003 Platform SDK anyway, which doesn't support anything older than Windows 95, and Windows NT before 3.5 didn't support WinSock at all.

_Side note:_ this means we can also get rid of some `LoadLibrary` calls in `telnet.c` and just call WinSock 2 directly:
https://github.com/curl/curl/blob/0f55269affc3cacd901af601b98faece2d62438d/lib/telnet.c#L1343-L1392

**Comment by ngg** [comment]
I tried to reproduce the slow sftp upload issue at https://github.com/curl/curl/issues/5633
I managed to reproduce it, but only with 7.71.0. For me it was working good when I tried with 7.71.1 (in contrast to what @rmja experienced there)
I also tried with this PR applied and it was still working good.

**Comment by mback2k** [comment]
@bagder Do you approve giving this another try for this release?

**Comment by jay** [comment]
If we can't reproduce what @rmja described I'd say go for it.

**Comment by bagder** [comment]
I agree with @jay. It would be good to get a test case added that would reproduce the previous bug and see that it no longer does...

**Comment by mback2k** [comment]
I will try to turn the example given by @rmja into a test case, to make sure we catch this error in the future.

**Comment by rmja** [comment]
Let me know if you can produce a prerelease build, then I can test to see if it works or not.

**Comment by mback2k** [comment]
Regarding the test case I wanted to create: we already have many FTP and SFTP upload test cases in our testsuite, so the use case should already be covered. The only problem is: we are currently not running SCP/SFTP tests on any Windows CI since I migrated my builds from buildbot to Azure Pipelines. I lost libssh2 support on the way and it is still on my ToDo list to bring that back.

In the mean time I would like to ask @rmja and @RadMission to check if this PR still has the problem for them. The Windows CI jobs for this PR successfully ran all the relevant HTTP and FTP upload tests at least, so if there is still any problem, I guess it is with SCP/SFTP. Please help us test this PR manually for now. Thanks in advance!

**Comment by rmja** [comment]
I will be happy to test this on Windows if you can send a link to a downloadable build.

**Comment by mback2k** [comment]
Okay, I can probably provide a debug build tomorrow.

**Comment by mback2k** [comment]
@rmja attached you will find a [debug build](https://github.com/curl/curl/files/4961739/curl-5634-20200722.zip) created with mingw-w64 from this PR on current [curl](https://github.com/curl/curl/commit/3cd9bc9689f8ddee23735cd1fbe90ce5c139ad02) and [libssh2](https://github.com/libssh2/libssh2/commit/5eaa7aa1c042e06cd4826a79af6af73ea5bf8370) master.

**Comment by rmja** [comment]
@mback2k My test suite fails with this build - I get SSL_CONNECT_ERROR when I do easy_perform. The curl builds that I have used until now has openssl and libcrypto as dependencies. I have of cause extracted all the files in the zip you provided, but is there any other dependency that I need to include for this build to work correctly?

**Comment by mback2k** [comment]
@rmja that build is just using libssh2 (with Windows-native WinCNG crypto backend) and curl (with Windows-native Schannel TLS backend). So it should work on recent Windows (>= Vista) without any additional dependencies except this mingw-w64 `libwinpthread` library.

**Comment by rmja** [comment]
Here is the outline of handle configuration. I have to proxy as the target server is IP filtered.

```
var handle = SafeEasyHandle.Init();

CurlNative.Easy.SetOpt(handle, CURLoption.USERNAME, credentials.UserName);
CurlNative.Easy.SetOpt(handle, CURLoption.PASSWORD, credentials.Password);
CurlNative.Easy.SetOpt(handle, CURLoption.PROXYTYPE, CurlProxyType.Socks5);
CurlNative.Easy.SetOpt(handle, CURLoption.PROXY, proxy.Host);
CurlNative.Easy.SetOpt(handle, CURLoption.PROXYPORT, proxy.Port);
CurlNative.Easy.SetOpt(handle, CURLoption.PROXYUSERPWD, proxy.UserInfo);
CurlNative.Easy.SetOpt(handle, CURLoption.SSL_VERIFYPEER, 0);
CurlNative.Easy.SetOpt(handle, CURLoption.URL, url.ToString());
CurlNative.Easy.SetOpt(handle, CURLoption.USE_SSL, CurlUseSsl.All);
CurlNative.Easy.SetOpt(handle, CURLoption.CUSTOMREQUEST, "MLSD");
CurlNative.Easy.SetOpt(handle, CURLoption.WRITEFUNCTION, (data, size, nmemb, user) =>
{
    // Handler is not called
});

var result = CurlNative.Easy.Perform(handle);
result == SSL_CONNECT_ERROR
```

**Comment by mback2k** [comment]
Ok, what type of protocol are you using? I thought you were using SFTP (via SSH), but it seems like you are using something based upon SSL, like HTTPS or FTPS? Can you give the (redacted) URL you are trying to use?

**Comment by rmja** [comment]
Yes, ftps. The url is ftps://some-server.com at port 990.

**Comment by mback2k** [comment]
Since the error `SSL_CONNECT_ERROR` is unrelated to this PR, can you try to connect to any public FTP server maybe?

**Comment by rmja** [comment]
Yes, connecting to public ftp works fine - that is also part of the test suite, and it passes fast!

**Comment by mback2k** [comment]
Is that also via SOCKS proxy? Please try more combinations and find out if just SSL is the problem now, if so, I could create another build using OpenSSL for you later.

**Comment by rmja** [comment]
Yes, that is also via a socks proxy - it is plain ftp on port 21 without any encryption. All the tests that fails (about 9 tests) are because  of the SSL_CONNECT_ERROR. I think that I have seen that error before because it was unable to find libcrypto. These are the native dlls that are being used/are present (except the -old.dll one):

image

**Comment by mback2k** [comment]
Okay, [attached](https://github.com/curl/curl/files/4974028/curl-5634-20200724.zip) you will find the same debug build with OpenSSL enabled (libssl and libcrypto). Please test that one.

**Comment by rmja** [comment]
Yes, now all tests passes meaning that the ssl error is gone, but they now take way too long meaning that the regression is still present as far as I can tell. With this build the tests passes in 54 seconds. With the older curl version they pass in 14 seconds.

**Comment by mback2k** [comment]
I have a theory about that. Will get into that later.

**Comment by mback2k** [comment]
Okay, I modified the socket readyness pre-check handling for [this build](https://github.com/curl/curl/files/4978232/curl-5634-20200726.zip). Please try it.

**Comment by rmja** [comment]
The build produces the same result :-/ - it still passes in around 55 seconds.

**Comment by mback2k** [comment]
Putting this on hold for the next feature window, until #5633 has been successfully tracked down and solved.

**Comment by mback2k** [comment]
@Togtja would you mind joining this PR by testing the builds above to see if they also affect the performance for you?

**Comment by Togtja** [comment]
@mback2k When building from [this](https://github.com/curl/curl/tree/8805a4b2a97c99a6c87d0cb61e1d9bc478b0fcd2) using CMake with vs build tools 2019, the performance seems to be exactly the same as `7.71.1` on downloads and uploads to the SFTP server. I am going to test with a few more variations of workload, but I suspect it's going to be the same.

While this PR and `7.71.1` is much faster to upload than `7.71.0`, it is still a fair bit slower than [Filezilla](https://filezilla-project.org/) that I normally use. curl uses about twice the amount of time. For downloads, FileZilla is just a tad bit faster 10-20%

**Comment by mback2k** [comment]
Please be more exact. Is 7.71.1 as fast as 7.70.0 for you and therefore this PR does also not affect that negatively?

I don't think we want to compare with other tools like FilaZilla right now, we first want to make sure that curl is not impacted by this.

**Comment by Togtja** [comment]
Sorry, I did not compare my results to `7.70.0`, However I have done so now, and this PR, your pre-compiled binaries (20200726 and 20200722), `7.70.0` and `7.71.1` all take about 30 seconds +/- 2s to upload 119Mb. 

`7.71.0` on the other hand does this in about 1,5 hours.

The testing I did was running
`./curl.exe -k -T test_of_large_zip_upload.zip sftp://user:pass@hodt:port/Path/to/folder/test_of_large_zip_upload.zip`
I ran this command 3 times for each build I had. Note that the PR, `7.70.0`, `7.71.0` and `7.71.1` was built on my computer using CMake with libnssh2 and OpenSSL:
`curl 7.7x.x (Windows) libcurl/7.7x.x OpenSSL/1.1.1g libssh2/1.9.0_DEV` x depending if it was `7.70.0`, `7.71.0` or `7.71.1`

The comparison to FileZilla was just a side note.

**Comment by mback2k** [comment]
Okay, thank you very much. So I guess we can assume that we will be able to merge this during the next feature window.

**Comment by mback2k** [comment]
@Togtja thank you very much @rmja can you confirm this by doing the same tests, especially comparing against 7.70.0?

File: lib/multi.c [hunk 3]
```diff
@@ -1067,11 +1073,16 @@ static CURLMcode Curl_multi_wait(struct Curl_multi *multi,
...
   unsigned int curlfds;
-  bool ufds_malloc = FALSE;
   long timeout_internal;
   int retcode = 0;
+#ifndef USE_WINSOCK
   struct pollfd a_few_on_stack[NUM_POLLS_ON_STACK];
   struct pollfd *ufds = &a_few_on_stack[0];
+  bool ufds_malloc = FALSE;
+#else
+  int already_writable = 0;
```
Comment (bagder) [inline lib/multi.c:1083]: maybe just use a `bool` for this instead ?

File: lib/multi.c [hunk 4]
```diff
@@ -1145,22 +1163,45 @@ static CURLMcode Curl_multi_wait(struct Curl_multi *multi,
...
+          struct timeval timeout;
+          fd_set writefds;
+          timeout.tv_sec = 0;
+          timeout.tv_usec = 0;
+          FD_ZERO(&writefds);
+          FD_SET(sockbunch[i], &writefds);
+          if(select((int)sockbunch[i] + 1, NULL, &writefds, NULL,
+                    &timeout) == 1)
+            already_writable++;
+          mask |= FD_WRITE;
```
Comment (bagder) [inline lib/multi.c:1190]: maybe this code be made into a separate sub-function as the exact same logic is done again further down?

File: lib/multi.c (hunk 3 above)
Comment (mback2k) [reply lib/multi.c:1083]: @rcombs would you mind applying my patch above, since it also addresses this?

File: lib/multi.c (hunk 4 above)
Comment (mback2k) [reply lib/multi.c:1190]: @rcombs would you mind applying my patch above, since it also addresses this by using existing macros?

**Comment by mback2k** [comment]
> I cannot comment on the code as this uses lots of Windows magic that I don't know anything about.

I requested your review in order to get a general approval to do this change since it broke curl in the past. 😉 

> I think the least positive thing with this code is the extreme amount new `#ifdefs` it adds, since it really creates a maze. Unfortunately I can't say I have any immediate ideas on how to remedy that.

Yes, I also thought whether it would be better to just create a new Windows-specific wait function and use that instead. What do you think?

> If the code builds and doesn't cause regressions I'm fine with the change as I think it improves curl on Windows.

We are not sure about this yet, as we are waiting on feedback from @rmja once he is available again.

**Comment by mback2k** [comment]
Since @rmja's issue in #5633 seems unrelated to the previous or current version of this change, I think we can merge it during the next feature window. @rcombs what do you think about my patch above? I might just do a follow up PR once this PR has landed.

**Comment by rcombs** [comment]
Patch overall looks good, but are the `SOCKET_READABLE` checks necessary? Seems like that'd add extraneous syscalls that shouldn't be needed, if I'm reading the Winsock docs correctly.

**Comment by mback2k** [comment]
@rcombs I think they are due to the fact that FD_CLOSE is only signaled once for a socket and this way we can avoid waiting on an already closed socket by pre-checking it with select which transparently handles this for us (even on Windows).

**Comment by rcombs** [comment]
Alright, rebased and re-pushed with @mback2k's changes applied; think we should be good to go now.

**Comment by mback2k** [comment]
@rcombs thanks, I will merge this once the release has cooled of a little and we can be sure there is no bugfix release needed.

**Pull request closed without merging**